### PR TITLE
#2 - Handle skipped cols Joi conflict

### PR DESCRIPTION
**Pull request by wswoodruff**
- Introduce new return code: `MIGRATION_WITH_CONFLICT` to inform the user that they have a Joi schema prop that conflicts with a skipped, unsupported db column
- Introduce `schemaConflict` property on all `skippedCols` entries so a user can search for which columns are an issue

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -193,19 +194,14 @@ exports.genMigrationFile = (options, cb) => {
                         const template = Fs.readFileSync(__dirname + '/hbs/migrationFileTemplate.hbs').toString();
                         const compiler = internals.handlebars().compile(template);
                         const createdTables = tableDeltas.tableDelta.create.concat(tableDeltas.joinTableDelta.create);
-                        const sameColumn = (skipA, skipB) => skipA.tableName === skipB.tableName && skipA.column === skipB.column;
-                        const dedupedSkippedCols = skippedCols.filter((skip, i) => {
-
-                            return skippedCols.findIndex((s) => sameColumn(s, skip)) === i;
-                        });
```
Comment (wswoodruff) [inline lib/index.js]: This deduping can be removed now that we prevent join table skipped columns from being re-added (See https://github.com/BigRoomStudios/schwifty-migrate-diff/pull/9/files#diff-6d186b954a58d5bb740f73d84fe39073R394)

File: lib/index.js [hunk 2]
```diff
@@ -572,12 +587,29 @@ internals.getJoinTableInfo = (model, knex, cb) => {
                         skippedCols.push({
                             tableName,
                             column,
-                            type
+                            type,
+                            schemaConflict: false // Set as a default, may be overridden below
```
Comment (wswoodruff) [inline lib/index.js]: Setting a default here to be overridden later is meh, idk doesn't feel great but thoughts?

File: lib/index.js [hunk 3]
```diff
@@ -572,12 +587,29 @@ internals.getJoinTableInfo = (model, knex, cb) => {
...
                     return !err;
                 });
 
+            const joiKeysToStrip = skippedCols.map((sc) => sc.column);
+
+            // Update skippedCols to say there's a schemaConflict
+            const onKeyStripped = (keyName) => {
+
+                skippedCols.find((sc) => sc.column === keyName).schemaConflict = true;
+            };
```
Comment (wswoodruff) [inline lib/index.js]: Modifying this object here instead of dependency-injecting it, how do we feel about that?

**Comment by devinivy** [comment]
I've been looking at this for a little while and don't have lots of specific feedback yet.  But I do have some more general feedback.  I get confused by the two very different paths that join tables and other tables go down.  For example, join tables require skipped columns to be excluded by [recreating a joi schema without certain fields](https://github.com/BigRoomStudios/schwifty-migrate-diff/blob/251923195eb1b35808d4d967015f065245e04c5f/lib/index.js#L597-L611) which causes all sorts of complications around setting `schemaConflict` and actually constructing the joi schema.  But at the same time all models go through a process of [filtering their "compiler schema"](https://github.com/BigRoomStudios/schwifty-migrate-diff/blob/251923195eb1b35808d4d967015f065245e04c5f/lib/index.js#L394-L416) to do the same thing (which seems like a good approach).  This seems to be related to the duplicated join table info, too.  These might be issues—not 100% sure—but overall the theme for me is that i struggle to grasp the big picture!  Maybe you could break it down for me some time.

**Comment by wswoodruff** [comment]
Ok yeah! I can definitely give rethinking some of the structure a thought, the join tables though are pretty special lil beasts in this -- and I agree the 'flow' of how the internals work do seem somewhat scrambled about

File: lib/index.js (hunk 3 above)
Comment (mattboutet) [reply lib/index.js]: This whole thing generally doesn't sit well with me.  I was a little surprised it worked at all, but what happens if `find()` doesn't find anything?

Comment (wswoodruff) [reply lib/index.js]: Outside of context and having not written it, I can see why that find would look scary. In here only keys from `joiKeysToStrip` will be passed to `onKeyStripped`, and joiKeysToStrip is filled with column names from `skippedCols`

**Comment by mattboutet** [comment]
I'd also be interested in a breakdown if it happens.  There's a lot going on here, and I had a hard time getting my head around it.

**Comment by wswoodruff** [comment]
Ok howbout at some point I can pick a couple situations found in the tests and we can meet IRL to discuss the logic flow that produced it?

**Comment by wswoodruff** [comment]
@devinivy Is this PR good to merge and then do the opt-in npm tagged release?

**Comment by wswoodruff** [comment]
NOTE we are going to remove `MIGRATION_WITH_CONFLICT` as a returnCode in a future version

**Pull request merged**